### PR TITLE
Fix subtotal and total top border

### PR DESCRIPTION
**Pull request by BrianBorge**
## Before

screen shot 2018-10-30 at 3 05 55 pm

## After
screen shot 2018-10-30 at 4 08 20 pm

### CODE REVIEW

File: lib/data-table/table.rb [hunk 1]
```diff
@@ -252,10 +252,17 @@ def render_group_recursive(collection, index = 1, group_parent = nil, ancestors
     #############
     def render_totals
       html = '<tfoot>'
+
+      first_rendered_total = nil
+
       @total_calculations.each_with_index do |totals_row, index|
+        if first_rendered_total.nil? && !totals_row.empty?
+          first_rendered_total = 'first'
```
Comment (aharpervc) [inline lib/data-table/table.rb]: won't this render `first` for all total rows, since the variable is defined outside of the loop block? The test case would be, if you have more than one total calculation, that they don't all get the css class.

Comment (BrianBorge) [reply lib/data-table/table.rb]: It will render first for all rows but not because the variable is defined outside of the loop block but because I never set `first_rendered_total` back to nil.

**edit** - I see what you're saying :)

Comment (BrianBorge) [reply lib/data-table/table.rb]: Thanks for the test suggestion.  About to add that and then refactor what I have.

Comment (aharpervc) [reply lib/data-table/table.rb]: You need to reset because you track this state over the life of the render_totals method, but why is that necessary? Isn't knowing whether or not a particular total row is the first total row knowable from entirely within this block?

Comment (BrianBorge) [reply lib/data-table/table.rb]: Yea, you are right.

Comment (BrianBorge) [reply lib/data-table/table.rb]: I just noticed Travis is running ci every time I push.

Comment (BrianBorge) [reply lib/data-table/table.rb]: @aharpervc - This is ready to be merged into it's parent branch.  Do you have any objections?

**Comment by BrianBorge** [comment]
@aharpervc - This is ready to be merged into it's parent branch.  Do you have any objections?

File: lib/data-table/table.rb [hunk 2]
```diff
@@ -254,7 +254,7 @@ def render_totals
       html = '<tfoot>'
       @total_calculations.each_with_index do |totals_row, index|
         next if totals_row.nil?
-        
+
```
Comment (aharpervc) [inline lib/data-table/table.rb]: unnecessary change?

File: lib/data-table/table.rb [hunk 3]
```diff
@@ -284,10 +284,19 @@ def render_subtotals(group_header, _group_data = nil, ancestors = nil)
       html = ''
       path = ancestors.nil? ? [] : ancestors.dup
       path << group_header
+
+      first_rendered_subtotal = nil
+
       @subtotal_calculations[path].each_with_index do |group, index|
+        if first_rendered_subtotal.nil? && !group.empty?
+          first_rendered_subtotal = 'first'
+        else
```
Comment (aharpervc) [inline lib/data-table/table.rb]: This logic feels weird. What is the value of `first_rendered_subtotal` for the second subtotal row? The logic to reset to `nil` will only run if first_rendered_subtotal is still `nil`. Is there a test case for this?

Comment (BrianBorge) [reply lib/data-table/table.rb]: It's `nil` for any row that isn't the first rendered one.  I modified an existing test in this PR, see changes below this comment.

#### Example Data Table with 2 subtotal rows

```ruby
collection = [
  {:name => 'Luke Skywalker', :class => 'Jedi Knight', :world => 'Star Wars', :power_level => 50},
  {:name => 'Emporer Palpatine', :class => 'Sith Lord', :world => 'Star Wars', :power_level => 95},
  {:name => 'Mithrander', :class => 'Wizard', :world => 'Middle Earth', :power_level => 9001},
  {:name => 'Aragorn', :class => 'Ranger', :world => 'Middle Earth', :power_level => 80}
]

DataTable.render(collection) do |t|
  t.column :power_level, "Power Level"
  t.column :mana, "Mana"

  t.group_by :world

  # two subtotal rows
  t.subtotal :power_level, :avg, 10 do |sub|
    "subtotal avg: #{sub}"
  end
  t.subtotal :mana, :avg, 11 do |sub|
    "subtotal avg: #{sub}"
  end
end
```

#### Generated HTML

```html


<table id="" class="data_table " cellspacing="0" cellpadding="0">
   <caption></caption>
   <thead>
      <tr>
         <th class="power_level ">Power Level</th>
         <th class="mana ">Mana</th>
      </tr>
   </thead>
   <tbody class="star_wars">
      <tr class="group_header level_0">
         <th colspan="2">Star Wars</th>
      </tr>
      <tr class="row_0 ">
         <td class="power_level numeric">50</td>
         <td class="mana nilclass"></td>
      </tr>
      <tr class="row_1 alt ">
         <td class="power_level numeric">95</td>
         <td class="mana nilclass"></td>
      </tr>
      <!-- first rendered subtotal row, gets top border -->
      <tr class="subtotal index_10 first">
         <td class="power_level numeric">subtotal avg: 72.5</td>
         <td class="mana nilclass"></td>
      </tr>
      <!-- second rendered subtotalrow, does not get top border -->
      <tr class="subtotal index_11 ">
         <td class="power_level numeric"></td>
         <td class="mana nilclass">subtotal avg: 0.0</td>
      </tr>
   </tbody>
   <tbody class="middle_earth">
      <tr class="group_header level_0">
         <th colspan="2">Middle Earth</th>
      </tr>
      <tr class="row_0 ">
         <td class="power_level numeric">9001</td>
         <td class="mana nilclass"></td>
      </tr>
      <tr class="row_1 alt ">
         <td class="power_level numeric">80</td>
         <td class="mana nilclass"></td>
      </tr>
      <!-- first rendered subtotal row, gets top border -->
      <tr class="subtotal index_10 first">
         <td class="power_level numeric">subtotal avg: 4540.5</td>
         <td class="mana nilclass"></td>
      </tr>
      <!-- second rendered subtotalrow, does not get top border -->
      <tr class="subtotal index_11 ">
         <td class="power_level numeric"></td>
         <td class="mana nilclass">subtotal avg: 0.0</td>
      </tr>
   </tbody>
</table>


```

File: lib/data-table/table.rb (hunk 2 above)
Comment (BrianBorge) [reply lib/data-table/table.rb]: I was just linting.  Should I remove?

Comment (aharpervc) [reply lib/data-table/table.rb]: I typically prefer to make separate commits for linting-only changes, so maybe do that?

Comment (BrianBorge) [reply lib/data-table/table.rb]: Sure, I changed it back.

File: lib/data-table/table.rb (hunk 3 above)
Comment (aharpervc) [reply lib/data-table/table.rb]: Interesting. Well, it it works, it works. It's hard to deduce from this change that it would work though, but that's probably just because I'm not very familiar with this code base.


Is it possible to move this if statement below the `next if group.empty?` line below? Seems odd to do 2 group.empty checks back to back.

**Review by aharpervc — APPROVED** [review]
2 comments otherwise LGTM to merge this all into the other branch
